Skip phi assignment if one of the merged states has an uninitialised object [blocks: #35, #2574, #3486]

tautschnig opened the pull request:
With level-2 counters incremented on declaration and non-deterministic
initialisation upon allocation, the only remaining sources are pointer
dereferencing, where uninitialised objects necessarily refer to invalid objects.

This is a cleaner implementation of 369f077d2e.

This is another bit factored out from #35.

tautschnig committed:
Skip phi assignment if one of the merged states has an uninitialised object

With level-2 counters incremented on declaration and non-deterministic
initialisation upon allocation, the only remaining sources are pointer
dereferencing, where uninitialised objects necessarily refer to invalid objects.

This is a cleaner implementation of 369f077d2e. Removing only the code
introduced in 369f077d2e would yield a wrong result for
regression/cbmc/Local_out_of_scope3.

Fixes: #1630

smowton reviewed (approved):
Should rename the PR? Since #2168 has landed in the meantime this is really just a cleanup now I think?

tautschnig commented:
> Should rename the PR? Since #2168 has landed in the meantime this is really just a cleanup now I think?

Not entirely: the implementation in #2168 did not yet consider the case that neither branch had SSA level2 at zero, which would occur with uninitialized globals, hence the argc/argv/envp fix in this PR before doing the "cleanup" part.

smowton commented:
Aha, in that case can we have a test to exhibit the change here?

tautschnig commented:
@smowton Apologies for the delay. I have amended the commit message, adding: "Removing only the code introduced in 369f077d2e would yield a wrong result for regression/cbmc/Local_out_of_scope3." -- so the test is actually already in place.

smowton commented:
Cool, in that case lgtm

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: ec8c07ca147b68f8ccf16f78d81dcd7edf344cc3).

peterschrammel reviewed (approved):
A couple of clang-format complaints, otherwise good to go.

smowton commented:
@tautschnig is there a reason you dropped this? Looks good to go AFAICT. There were a couple of spurious Travis failures that I just restarted.

tautschnig commented:
@kroening ping?

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: cd59f1b68bca8f9d22f716a5261e6f665a3eb27c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86845275

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: a18f5d7e5d7164670a21727e1b34b901f7765ffe).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88753514

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2311826c696f2eb68579e3c790acc03e179061f5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90528381

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c7f9d6e9b9cfccdd8ea84109cc0c31a2775360b4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90998307

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8f1745deba0c2f570a86be75322c9872399e5640).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93272998

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 04d59575fe2c103f6990d3dd84d8fad36479c80f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93794605

tautschnig committed:
Fix handling of extern symbols in phi_function

In 646cf29941499, initialisation of dynamic objects was added, but extern
symbols were not considered. Includes a test that fails without this fix.

tautschnig commented:
This now fixes even more bugs with fewer changes.

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -469,17 +469,23 @@ static void merge_names(
+    (dest_count == 0 && !symbol.value.is_not_nil()))
   {
-    rhs = dest_state_rhs;
+    if(goto_count == 0)
```
Note the `dest_count == 0 && goto_count == 0` case is forbidden above, so this implies `dest_state.guard.is_false() && dest_count != 0) -- not sure if this can be distangled to make that clearer, or perhaps clarify with a comment

tautschnig replied:
If `dest_state.guard.is_false()` then really we don't care about `dest_count`, because it sits on a dead branch. Any idea what you'd like the comment to say?

smowton replied:
How about just pulling out

```
if(
  (dest_state.guard.is_false() && goto_count == 0) ||
  (goto_state.guard.is_false() && dest_count == 0))
{
  return;
}
```

It does mean testing `is_false()` multiple times, but it makes it clearer which cases are possible. Comment-wise, I guess we should add something like `// If the symbol is not initialised on one path and the other path is unreachable we don't need to generate anything`

tautschnig replied:
Mind taking another look at the latest version of this PR? Your comments helped make clear which change fixes which bug!

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 72f3bea802316de116f6c40171d47f78f64c72f8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95266652

smowton reviewed (approved):
Looks good! I have a feeling the previous version used `!is_not_nil()` where the current one uses `is_not_nil()` (without the `!`), so double-check that?

tautschnig commented:
> Looks good! I have a feeling the previous version used `!is_not_nil()` where the current one uses `is_not_nil()` (without the `!`), so double-check that?

Well, one of them did... - which was a bug...

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: eaa7d93ac4d9882149fb0952c7d1177ce1840236).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95298901
